File: jeehyun/공통문제/백준/18430_무기 공학.cpp
#include <iostream>
using namespace std;

int n, m;
int map[5][5];
bool visited[5][5];
int answer = 0;

int rightUp(int x, int y) // ㄱ
{
	return 2 * map[x][y] + map[x + 1][y] + map[x][y - 1];
}

int rightDown(int x, int y) // _ㅣ
{
	return 2 * map[x][y] + map[x - 1][y] + map[x][y - 1];
}

int leftUp(int x, int y) // r
{
	return 2 * map[x][y] + map[x + 1][y] + map[x][y + 1];
}

int leftDown(int x, int y) // ㄴ
{
	return 2 * map[x][y] + map[x - 1][y] + map[x][y + 1];
}

void dfs(int x, int y, int sum) 
{
	if (y == m) { // 다음줄로 넘어간다
		y = 0;
		x++;
	}

	if (x == n) { // 부메랑을 다 만든 경우
		answer = max(answer, sum);
		return;
	}

	if (!visited[x][y]) {
		if (x + 1 < n && y - 1 >= 0 && !visited[x + 1][y] && !visited[x][y - 1]) {
			visited[x][y] = true;
			visited[x + 1][y] = true;
			visited[x][y - 1] = true;
			int nsum = sum + rightUp(x, y);
			dfs(x, y + 1, nsum);
			visited[x][y] = false;
			visited[x + 1][y] = false;
			visited[x][y - 1] = false;
		}

		if (x - 1 >= 0 && y - 1 >= 0 && !visited[x - 1][y] && !visited[x][y - 1]) {
			visited[x][y] = true;
			visited[x - 1][y] = true;
			visited[x][y - 1] = true;
			int nsum = sum + rightDown(x, y);
			dfs(x, y + 1, nsum);
			visited[x][y] = false;
			visited[x - 1][y] = false;
			visited[x][y - 1] = false;
		}

		if (x + 1 < n && y + 1 < m && !visited[x + 1][y] && !visited[x][y + 1]) {
			visited[x][y] = true;
			visited[x + 1][y] = true;
			visited[x][y + 1] = true;
			int nsum = sum + leftUp(x, y);
			dfs(x, y + 1, nsum);
			visited[x][y] = false;
			visited[x + 1][y] = false;
			visited[x][y + 1] = false;
		}

		if (x - 1 >= 0 && y + 1 < m && !visited[x - 1][y] && !visited[x][y + 1]) {
			visited[x][y] = true;
			visited[x - 1][y] = true;
			visited[x][y + 1] = true;
			int nsum = sum + leftDown(x, y);
			dfs(x, y + 1, nsum);
			visited[x][y] = false;
			visited[x - 1][y] = false;
			visited[x][y + 1] = false;
		}
	}

	// 부메랑을 만들지 않는다
	dfs(x, y + 1, sum);
}

int main()
{
	cin >> n >> m;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> map[i][j];
		}
	}

	dfs(0, 0, 0);
	cout << answer;
	
	return 0;
}

File: seojin/개별문제/0501/18430_무기공학.py
# 너무 복잡해...알고싶지않다.....
def rightUp(y,x):
  global namu
  return namu[y][x-1]+namu[y+1][x]+2*namu[y][x]

def rightDown(y,x):
  global namu
  return namu[y][x-1]+namu[y-1][x]+2*namu[y][x]

def leftUp(y,x):
  global namu
  return namu[y][x+1]+namu[y+1][x]+2*namu[y][x]

def leftDown(y,x):
  global namu
  return namu[y][x+1]+namu[y-1][x]+2*namu[y][x]

ans = 0
def dfs(y, x, sum): 
  global n, m, visited, ans
  if x == m:
    x = 0
    y+=1
  if y == n:
    ans = max(ans,sum)
    return
  if visited[y][x]!=1:
    if (y+1<n) and (x-1>=0) and (visited[y+1][x]!=1) and (visited[y][x-1]!=1): # 경우 1(ㄱ 모양)
      visited[y][x] = 1
      visited[y+1][x] = 1
      visited[y][x-1] = 1
      nsum = sum+rightUp(y,x);
      dfs(y,x+1,nsum)
      # 끝까지 갔다왔으면 다시 되돌려놔야함
      visited[y][x] = 0 
      visited[y+1][x] = 0
      visited[y][x-1] = 0
    if (y-1>=0) and (x-1>=0) and (visited[y-1][x]!=1) and (visited[y][x-1]!=1): # 경우 2(ㄴ 뒤집은 모양)
      visited[y][x] = 1
      visited[y-1][x] = 1
      visited[y][x-1] = 1
      nsum = sum+rightDown(y,x);
      dfs(y,x+1,nsum)
      # 끝까지 갔다왔으면 다시 되돌려놔야함
      visited[y][x] = 0 
      visited[y-1][x] = 0
      visited[y][x-1] = 0
    if (y+1<n) and (x+1<m) and (visited[y+1][x]!=1) and (visited[y][x+1]!=1): # 경우 3(ㄱ 뒤집은 모양)
      visited[y][x] = 1
      visited[y+1][x] = 1
      visited[y][x+1] = 1
      nsum = sum+leftUp(y,x);
      dfs(y,x+1,nsum)
      # 끝까지 갔다왔으면 다시 되돌려놔야함
      visited[y][x] = 0 
      visited[y+1][x] = 0
      visited[y][x+1] = 0
    if (y-1>=0) and (x+1<m) and (visited[y-1][x]!=1) and (visited[y][x+1]!=1): # 경우 3(ㄱ 뒤집은 모양)
      visited[y][x] = 1
      visited[y-1][x] = 1
      visited[y][x+1] = 1
      nsum = sum+leftDown(y,x);
      dfs(y,x+1,nsum)
      # 끝까지 갔다왔으면 다시 되돌려놔야함
      visited[y][x] = 0 
      visited[y-1][x] = 0
      visited[y][x+1] = 0
  dfs(y,x+1,sum)

n, m = map(int, input().split())
namu = []
for i in range(n):
  namu.append(list(map(int, input().split())))
visited = [[0]*m for i in range(n)]

dfs(0,0,0)
print(ans)